File: Taller3/crecRegiones-ITK/C++/ConnectedThreshold.cxx
#include "itkImageFileReader.h"
#include "itkConnectedThresholdImageFilter.h"
#include "itkRescaleIntensityImageFilter.h"
#include "itkImageFileWriter.h"

int
main(int argc, char * argv[])
{
  // sample usage
  //./connectedThreshold input output 100 170 132 142 96

  if (argc < 8)
  {
    std::cerr << "Usage: " << std::endl;
    std::cerr << argv[0] << std::endl;
    std::cerr << " <InputImage> <OutputImage> <LowerThreshold>";
    std::cerr << " <UpperThreshold> <XSeed> <YSeed> <ZSeed>";
    return EXIT_FAILURE;
  }

  constexpr unsigned int Dimension = 3;
  using PixelType = short;
  using SizeType = itk::SizeValueType;

  const char * InputImage = argv[1];
  const char * OutputImage = argv[2];

  const auto LowerThreshold = static_cast<SizeType>(atoi(argv[3]));
  const auto UpperThreshold = static_cast<SizeType>(atoi(argv[4]));

  const auto XSeed = static_cast<SizeType>(atoi(argv[5]));
  const auto YSeed = static_cast<SizeType>(atoi(argv[6]));
  const auto ZSeed = static_cast<SizeType>(atoi(argv[7]));

  using ImageType = itk::Image<PixelType, Dimension>;

  using ReaderType = itk::ImageFileReader<ImageType>;
  ReaderType::Pointer reader = ReaderType::New();
  reader->SetFileName(InputImage);
  reader->Update();

  ImageType::Pointer image = reader->GetOutput();
  ImageType::RegionType region = image->GetLargestPossibleRegion();
  ImageType::SizeType size = region.GetSize();
  std::cout << size << std::endl;

  using FilterType = itk::ConnectedThresholdImageFilter<ImageType, ImageType>;
  FilterType::Pointer connectedThreshold = FilterType::New();
  connectedThreshold->SetLower(LowerThreshold);
  connectedThreshold->SetUpper(UpperThreshold);
  connectedThreshold->SetReplaceValue(255);

  ImageType::IndexType seed;
  seed[0] = XSeed;
  seed[1] = YSeed;
  seed[2] = ZSeed;
  connectedThreshold->SetSeed(seed);
  connectedThreshold->SetInput(reader->GetOutput());

  using RescaleType = itk::RescaleIntensityImageFilter<ImageType, ImageType>;
  RescaleType::Pointer rescaler = RescaleType::New();
  rescaler->SetInput(connectedThreshold->GetOutput());
  rescaler->SetOutputMinimum(0);
  rescaler->SetOutputMaximum(255);

  using WriterType = itk::ImageFileWriter<ImageType>;
  WriterType::Pointer writer = WriterType::New();
  writer->SetFileName(OutputImage);
  writer->SetInput(rescaler->GetOutput());

  try
  {
    writer->Update();
  }
  catch (itk::ExceptionObject & e)
  {
    std::cerr << "Error: " << e << std::endl;
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}

File: Taller3/crecRegiones-ITK/Python/ConnectedThreshold.py
import itk
import sys

# sample usage
# ./connectedThreshold input output 100 170 132 142 96

if len(sys.argv) < 8 :
  print("Usage: ", sys.argv[0], " <InputImage> <OutputImage> <LowerThreshold> <UpperThreshold> <XSeed> <YSeed> <ZSeed>")
  sys.exit()

Dimension = 3
PixelType = itk.US

InputImage = sys.argv[1]
OutputImage = sys.argv[2]

LowerThreshold = int(sys.argv[3])
UpperThreshold = int(sys.argv[4])

XSeed = int(sys.argv[5])
YSeed = int(sys.argv[6])
ZSeed = int(sys.argv[7])

ImageType = itk.Image[PixelType, Dimension]

ReaderType = itk.ImageFileReader[ImageType]
reader = ReaderType.New()
reader.SetFileName(InputImage)
reader.Update()

image = reader.GetOutput()
region = image.GetLargestPossibleRegion()
size = region.GetSize()
print(size)

FilterType = itk.ConnectedThresholdImageFilter[ImageType, ImageType]
connectedThreshold = FilterType.New()
connectedThreshold.SetLower(LowerThreshold)
connectedThreshold.SetUpper(UpperThreshold)
connectedThreshold.SetReplaceValue(255)

seed = []
seed.append(XSeed)
seed.append(YSeed)
seed.append(ZSeed)
connectedThreshold.SetSeed(seed)
connectedThreshold.SetInput(reader.GetOutput())

RescaleType = itk.RescaleIntensityImageFilter[ImageType, ImageType]
rescaler = RescaleType.New()
rescaler.SetInput(connectedThreshold.GetOutput())
rescaler.SetOutputMinimum(0)
rescaler.SetOutputMaximum(255)

WriterType = itk.ImageFileWriter[ImageType]
writer = WriterType.New()
writer.SetFileName(OutputImage)
writer.SetInput(rescaler.GetOutput())
writer.Update()
